### PR TITLE
Selectors API: Make duotone selectors fallback and be scoped

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related:

- https://github.com/WordPress/gutenberg/pull/46496
- https://github.com/WordPress/gutenberg/pull/49393

## What?

This PR fixes global-level duotone (see https://github.com/WordPress/gutenberg/pull/49393#issuecomment-1488425345 for details) and make duotone work like any other selector.

## Why?

- Make every selector work the same.
- Fix global-level duotone.

## How?

- Make duotone selectors work like any other selector:
  - blocks provide the whole selector via its block.json
  - it fall backs to the root block selector if requested

- Updates the selectors API functions to scope the old `supports.color.__experimentalDuotone` selector automatically. 
- Stabilize duotone support flag under `filter.duotone`
  - Leverages `block_type_metadata_settings` filter to migrate old `color.__experimentalDuotone` flag.

## Test

- Create a post that contains two images (uses `selectors.filter.duotone`) and two cover blocks (`supports.color.__experimentalDuotone`).
- Block level duotone: set the `purple-green` duotone for one of the images and one of the covers.
- Global level duotone: go to the site editor > global styles > blocks. Set the midnight duotone for the image and cover blocks.

The expected result is that the blocks with block-level duotone have it applied and the other ones use the global-level duotone in any context (front-end, post editor, site editor).

No other `img` elements of the page have any duotone applied (check the user avatar at the top-right, for example).

Front-end:

image

Post editor:

image

### CODE REVIEW

File: lib/compat/wordpress-6.3/get-global-styles-and-settings.php
Comment (scruffian) [inline lib/compat/wordpress-6.3/get-global-styles-and-settings.php]: Is it worth saying that this is left in for backwards compatibility, and should be removed one day?

File: lib/class-wp-duotone-gutenberg.php
Comment (oandregal) [inline lib/class-wp-duotone-gutenberg.php]: I wish I hadn't to paste this here, but this needs to work differently than the `WP_Theme_JSON_Gutenberg::scope_selector` (do not add an extra space).

Ideas for a follow-up PR:

- The usage of this internal utility has expanded to: [settings](https://github.com/WordPress/gutenberg/blob/70198b9811c1bdc99579e272a72af3dae0ec9138/lib/block-supports/settings.php#L95), [duotone](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-duotone-gutenberg.php#L352), [WP_Theme_JSON class](https://github.com/WordPress/gutenberg/blob/70198b9811c1bdc99579e272a72af3dae0ec9138/lib/class-wp-theme-json-gutenberg.php#L998). It's worth creating a utility function somewhere. Where? `wp_scope_css_selector( $scope, $selector ) ` as a name?

- The consumer should be responsible for adding or not adding spaces, not the function itself:
  - `wp_scope_css_selector( $scope . ' ', $selector )` if the consumer wants spaces.
  - `wp_scope_css_selector( $scope . ' ', $selector )` if the consumer doesn't want spaces.

Thoughts?

File: lib/class-wp-theme-json-gutenberg.php
Comment (oandregal) [inline lib/class-wp-theme-json-gutenberg.php:2403]: Duotone is already scoped in `get_blocks_metadata` by using `wp_get_block_css_selector`.

**Review by oandregal — APPROVED** [review]
This is working as expected.

File: packages/block-editor/src/hooks/duotone.js
Comment (scruffian) [inline packages/block-editor/src/hooks/duotone.js:242]: I think it probably makes more sense to keep this function but just modify it.

File: packages/block-editor/src/components/global-styles/use-global-styles-output.js
Comment (ajlende) [inline packages/block-editor/src/components/global-styles/use-global-styles-output.js:865]: Scoping the selector here was moved [below](https://github.com/WordPress/gutenberg/pull/49423/files#diff-52acb55f718dce035b748387f60ef068cd3769674b553584f7df6f790e9fa84eR1011-R1021) in `getBlockSelectors` where it was easier.

File: packages/block-editor/src/hooks/duotone.js
Comment (ajlende) [inline packages/block-editor/src/hooks/duotone.js:232]: I did a lot of refactoring here so the JS could more closely match `WP_Duotone_Gutenberg::render_duotone_support`

File: packages/block-editor/src/hooks/duotone.js
Comment (ajlende) [inline packages/block-editor/src/hooks/duotone.js]: I figured this was worth memoizing because there were some performance issues noted in #48401 and I had to move some things back in here.

File: lib/class-wp-duotone-gutenberg.php
Comment (aaronrobertshaw) [inline lib/class-wp-duotone-gutenberg.php]: We might need to document that any duotone selector must begin with the block's class as there isn't anything in terms of code to enforce that.

**Review by aaronrobertshaw — COMMENTED** [review]
Thank you very much for all the hard work here! 

I found while testing that the stabilization of the duotone support flag under `filter.duotone` caused a number of bugs. The main ones being that any blocks using `filter.duotone` instead of `color.__experimentalDuotone` no longer got the duotone control in the block toolbar or the filters screen in global styles.

Screenshot 2023-03-30 at 2 34 01 pm


Trying to support both flags here was a little tricky. The best I came up with was leveraging the `block_type_metadata_settings` to use the `color.__experimentalDuotone` flag to set `filter.duotone`. 

With that in place we could consolidate the block support checks and reinstate the duotone controls for all supported blocks. I think I've updated all the locations we check for support but perhaps a double-check by those more familiar with duotone support would be prudent.

If using the `block_type_metadata_settings` filter is acceptable, I think another follow-up could be to use it to migrate all `__experimentalSelector` properties into the `selectors` property. That would in turn allow us to simplify the Selectors API code further.

Other than that, I made a few other minor edits:
- Updated block support docs now we have a non-experimental duotone support flag
- Added `filter.duotone` to block.json schema
- Fixed other minor typos and nits

This is testing well for me now:
✅ Global duotone filters apply in all editors and frontend
✅ Individual duotone filters apply in all editors and frontend
✅ Duotone controls display when they should

It would be great to get a final review given the changes made today. If no further issues are encountered, please feel free to merge this one. Otherwise, I'll merge it tomorrow.

File: lib/block-supports/duotone.php
Comment (scruffian) [inline lib/block-supports/duotone.php]: This looks like a useful technique for migrating more settings over.

**Review by scruffian — APPROVED** [review]
This seems to be working as expected to me, and the code looks good, but I'd rather wait on @ajlende to merge it.

File: docs/reference-guides/block-api/block-supports.md
Comment (ajlende) [inline docs/reference-guides/block-api/block-supports.md:232]: Seeing this removed makes me so happy 😁

File: lib/class-wp-duotone-gutenberg.php
Comment (ajlende) [reply lib/class-wp-duotone-gutenberg.php]: Technically this would work with anything that isn't a type selector (which includes tag names and namespaces). You can work around the tag names with the `:is()` pseudo-class, but namespaces are harder.

File: lib/block-supports/duotone.php
Comment (ajlende) [reply lib/block-supports/duotone.php]: The only problem is that there's no longer any way of knowing if you have experimental support or not. It's causing me problems because the selector that we choose is different depending on whether or not it's experimental.

```js
// May be true even if `filter.duotone` isn't set.
const duotoneSupport = getBlockSupport(
	blockType,
	'filter.duotone',
	false
);

// May be set even if `filter.duotone` is also set.
const experimentalDuotone = getBlockSupport(
	blockType,
	'color.__experimentalDuotone',
	false
);

// Will always produce a selector even if `filter` or `filter.duotone` isn't set.
const duotoneSelector = getBlockCSSSelector(
	blockType,
	'filter.duotone',
	{ fallback: true }
);
```

EDIT: We could have `color.__experimentalDuotone` override `filter.duotone`, but that seems wrong to me. If they're both set, I would expect `filter.duotone` to override `color.__experimentalDuotone`.

File: packages/block-editor/src/hooks/duotone.js
Comment (ajlende) [inline packages/block-editor/src/hooks/duotone.js]: I had to do this to fix the experimental selectors in the editor being applied to the root of the block instead of the selector set in __experimentalSelector.

**Review by ajlende — APPROVED** [review]
@aaronrobertshaw The changes I made ended up being kind of big again, so I didn't want to merge right away without someone else taking a look. I have nothing else to add, so if things are working well for you, go ahead and merge it.

File: lib/block-supports/duotone.php
Comment (aaronrobertshaw) [reply lib/block-supports/duotone.php]: I'm not sure I follow.

In the filter, we were only setting `filter.duotone` if it hasn't already been set and the `color.experimentalDuotone` has been.

A block author shouldn't be using both properties in their raw block.json so if we have `color.experimentalDuotone` set at all, then it should be a case from when duotone was experimental.

- In your first example, the check for `filter.duotone` tells us if there is support.
- The second example if the default is removed or changed to `null` would tell us if the duotone support was the experimental version.
- The third snippet with fallback true would always produce a selector regardless of any support check. We would guard such a call with a check that the `color.experimentalDuotone` support flag wasn't set.

Comment (aaronrobertshaw) [reply lib/block-supports/duotone.php]: My thinking was that by leaving the `__experimentalDuotone` property we could determine when to use that as the selector. I can see from this discussion where I missed adding the guard check to the duotone.js HoC.

File: packages/block-editor/src/hooks/duotone.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/duotone.js]: I'm not sure we're on the same page here. The `supports.filter.duotone` property will only get set if it hasn't been set already and we have `supports.color.__experimentalDuotone` set.

So `supports.filter.duotone` will tell us if we have duotone support one way or another. Checking if we have the `supports.color.__experimentalDuotone` flag set should tell us if we need to handle an experimental duotone selector or use the selectors API requesting `selectors.filter.duotone`

**Comment by aaronrobertshaw** [comment]
I've updated the comments added around backwards compatibility of `supports.color.__experimentalDuotone` to match the filter that only updates `supports.filter.duotone` if it wasn't already set. 

We can also return earlier if there is no duotone support so I relocated those guard conditions as well in https://github.com/WordPress/gutenberg/pull/49423/commits/3346b9aaf4fd250de1158076820bb857bb74fe15

I've retested this:

✅ Global styles can be applied to all current blocks opting into `__experimentalDuotone`
✅ Global styles are applied correctly for the Image block that uses `supports.filter.duotone`
✅ Individual block styles in the editor can be applied for blocks using `__experimentalDuotone`
✅ Individual block styles work for the Image block using `supports.filter.duotone`
✅ Duotone filters via global styles apply for all blocks on frontend
✅ Individual block duotone filters override global duotone filters on the frontend

I'll get some fresh eyes on this PR and look to merge it later this afternoon.

**Review by glendaviesnz — APPROVED** [review]
I have retested after your latest changes @aaronrobertshaw and the following still works for me:

✅ Global styles can be applied to all current blocks opting into __experimentalDuotone
✅ Global styles are applied correctly for the Image block that uses supports.filter.duotone
✅ Individual block styles in the editor can be applied for blocks using __experimentalDuotone
✅ Individual block styles work for the Image block using supports.filter.duotone
✅ Duotone filters via global styles apply for all blocks on frontend
✅ Individual block duotone filters override global duotone filters on the frontend

**Comment by ryanwelcher** [comment]
I just cherry-picked this PR to the release/15.5 branch to get it included in the next release: fa6e2a6620

**Pull request merged**
